pyemma/coordinates/clustering/kmeans.py

```diff
@@ -193,3 +185,107 @@ def _param_add_data(self, X, itraj, t, first_chunk, last_chunk_in_traj, last_chu
+                                                    self.metric, self.n_clusters)
+                self._cluster_centers = [c for c in cc]
+
+    def _collect_data(self, X, first_chunk, stride):
+        # beginning - compute
+        if first_chunk:
+            self._t_total = 0
+            mem_req = int(1.0 / 1024 ** 2 * X[0, :].nbytes * self.n_frames_total(stride))
+            if mem_req > 10:
+                self._logger.warn('K-means implementation is currently memory inefficient.'
+                                  ' This calculation needs %i megabytes of main memory.'
+                                  ' If you get a memory error, try using a larger stride.'
+                                  % mem_req)
+
+        # appends a true copy
+        self._in_memory_chunks[self._t_total:self._t_total + len(X)] = X[:]
+        self._t_total += len(X)
+
+
+class MiniBatchKmeansClustering(KmeansClustering):
+    def __init__(self, n_clusters, max_iter=5, metric='euclidean', tolerance=1e-5, init_strategy='kmeans++',
+                 batch_size=0.2, oom_strategy='memmap'):
+        super(MiniBatchKmeansClustering, self).__init__(n_clusters, max_iter, metric, tolerance, init_strategy,
+                                                        oom_strategy)
+        self._batch_size = batch_size
+        if self._batch_size > 1:
+            raise ValueError("batch_size should be less or equal to 1, but was %s" % batch_size)
+
+    def _init_in_memory_chunks(self, size):
+        return super(MiniBatchKmeansClustering, self)._init_in_memory_chunks(self._n_samples)
+
+    def _draw_mini_batch_sample(self):
+        offset = 0
+        for idx, traj_len in enumerate(self._traj_lengths):
+            self._random_access_stride[offset:offset + self._n_samples_traj[idx], 0] = idx * np.ones(
+                self._n_samples_traj[idx], dtype=int)
+            self._random_access_stride[offset:offset + self._n_samples_traj[idx], 1] = np.sort(
+                random.sample(xrange(traj_len), self._n_samples_traj[idx])).T
+            offset += self._n_samples_traj[idx]
+        return self._random_access_stride
+
+    def _param_init(self):
+        self._traj_lengths = self.trajectory_lengths(stride=self._param_with_stride)
+        self._total_length = sum(self._traj_lengths)
+        samples = int(math.ceil(self._total_length * self._batch_size))
+        self._n_samples = 0
+        self._n_samples_traj = {}
+        self._prev_cost = 0
+        for idx, traj_len in enumerate(self._traj_lengths):
+            traj_samples = int(math.floor(traj_len / float(self._total_length) * samples))
+            self._n_samples_traj[idx] = traj_samples
+            self._n_samples += traj_samples
+
+        self._random_access_stride = np.empty(shape=(self._n_samples, 2), dtype=int)
+
+        super(MiniBatchKmeansClustering, self)._param_init()
+
+        return 0, self._draw_mini_batch_sample()
```

Comment:
Actually, i have - in transformer.py lines 331 up to 339. This is necessary because i want to have a sampled version of the data before _param_add_data gets called for the first time.